// adapter/Errors.js
var Errors = {
    unknown             : {code: -1, msg: 'Не известная ошибка', mail : true},
    inner               : {code: -2, msg: 'Используем текущую'},
    common : {
        create_dir      : {code: 1020, msg: 'Ошибка создания директории'},
        write_file      : {code: 1021, msg: 'Ошибка записи в файл'}
    },
    soap : {
        empty_request   : {code: 1101, msg: 'Нельзя отправить пустой запрос адаптеру'},
        empty_response  : {code: 1102, msg: 'Пришел пустой ответ'},
        communication   : {code: 1103, msg: 'Ошибка соединения с адаптером'},
        status          : {code: 1104, msg: 'Плохой ответ от адаптера'}
    },
    xml : {
        to_object       : {code: 1201, msg: 'Ошибка преобразования XML'},
        from_object     : {code: 1202, msg: 'Ошибка преобразования объекта'},
        empty_xml       : {code: 1203, msg: 'Пустая Xml'}
    },
    database : {
        connect         : {code: 1301, msg: 'Ошибка подключения к базе данных', mail : true},
        query           : {code: 1302, msg: 'Ошибка запроса к базе данных', mail : true},
        empty           : {code: 1303, msg: 'В БД нет данных по запросу', mail : true},
        reconnect       : {code: 1304, msg: 'Ошибка переподключения к базе данных'},
        skipped         : {code: 1305, msg: 'Запрос пропущен через предыдущие ошибки'},
        set_encoding    : {code: 1306, msg: 'Ошибка установки кодировки'},
        bad_array       : {code: 1307, msg: 'Пустой массив'},
        query_timeout   : {code: 1308, msg: 'Таймаут'}
    },
    cache : {
        connect         : {code: 1401, msg: 'Ошибка подключения к кэш серверу', mail : true},
        wrong_cell      : {code: 1402, msg: 'Плохой индекс ячейки кэша'},
        read            : {code: 1403, msg: 'Ошибка получения значения ячейки кэша'},
        empty           : {code: 1404, msg: 'Пустая ячейка кэша'},
        write           : {code: 1405, msg: 'Ошибка записи в кэш'},
        not_writed      : {code: 1406, msg: 'Не удалось записать в кэш'},
        wrong_data      : {code: 1407, msg: 'Плохие данные для записи в кэш'},
        life_time       : {code: 1408, msg: 'Не верное время жизни данных в ячейке'},
        bad_array       : {code: 1409, msg: 'Плохой массив для записи в кэш'},
        write_array     : {code: 1410, msg: 'Не все елементы записаны в кэш'}
    },
    server : {
        no_worker       : {code: 1501, msg: 'Определите "worker" для сервера'},
        error           : {code: 1502, msg: 'Ошибка сервера'},
        empty_request   : {code: 1503, msg: 'Ошибка сервера: пустой запрос к серверу'},
        disconnected    : {code: 1504, msg: 'Ошибка сервера: клиент отсоединился'},
        bad_sid         : {code: 1505, msg: 'Ошибка сервера: сессия не существует'},
        uncaught_expt   : {code: 1506, msg: 'Ошибка сервера: внутренняя ошибка'} // ошибка кода
    }
};

module.exports = Errors;
